.github/workflows/publish.yml

```diff
@@ -7,18 +7,32 @@ on:
 jobs:
   publish:
     runs-on: ubuntu-latest
+    defaults:
+      run:
+        shell: bash
+    environment:
+      name: pypi
+      url: https://pypi.org/project/aiomqtt
+    permissions:
+      id-token: write
     steps:
-      - name: Checkout
+      - name: Checkout repository
         uses: actions/checkout@v3
+        with:
+          fetch-depth: 0 # Fetch entire history for poetry-dynamic-versioning, see: https://github.com/mtkennerly/poetry-dynamic-versioning/issues/55
+      - name: Install Poetry
+        uses: snok/install-poetry@v1
+        with:
+          virtualenvs-create: true
+          virtualenvs-in-project: true
+          installer-parallel: true
       - name: Set up Python
-        uses: actions/setup-python@v4
+        uses: actions/setup-python@v4 # Uses the Python version in .python-version
         with:
-          python-version: 3.x
-      - name: Install dependencies
-        run: pip install build
-      - name: Create package
-        run: python -m build -s -w .
-      - name: Upload package
+          cache: "poetry"
+      - name: Setup dynamic versioning
+        run: poetry self add "poetry-dynamic-versioning[plugin]"
+      - name: Build package
+        run: poetry build
```

Comment:
@empicano FYI, building in the same job as publishing is insecure. The build toolchain mustn't have access to OIDC privilege: https://github.com/marketplace/actions/pypi-publish#non-goals.